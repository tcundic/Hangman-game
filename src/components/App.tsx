import React from 'react';
import {connect} from 'react-redux';
import {Navigate} from "react-router-dom";
import Props from "../models/Props";
import {setUserName} from "../actions/games";
import {mapDispatchToProps, mapStateToProps} from "../utils/utilMethods";

export class App extends React.Component<Props> {
    state = {
        username: ''
    };

    onUsernameSave = (): void => {
        this.props.dispatch(setUserName(this.state.username));
    };

    onUsernameChange = (e: React.ChangeEvent<any>): void => {
        const username = e.target.value;
        this.setState(() => ({username}));
    }

    render() {
        if (this.props.game?.userName) {
            return <Navigate to="/play" />;
        }

        return (
            <div className="container mt-6">
                <label className="label is-large">Username</label>
                <div className="field is-grouped">
                    <p className="control">
                        <input
                            className="input is-large"
                            type="text"
                            placeholder="Enter username"
                            value={this.state.username}
                            onChange={this.onUsernameChange}
                        />
                    </p>
                    <p className="control">
                        <button className="button is-info is-large" onClick={this.onUsernameSave}>
                            Save username
                        </button>
                    </p>
                </div>
            </div>
        );
    }
};

export default connect(mapStateToProps, mapDispatchToProps)(App);