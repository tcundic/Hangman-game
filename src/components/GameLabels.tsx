import React from "react";
import Props from "../models/Props";
import Game from "../models/game";
import Highscore from "../models/highscore";
import {connect} from "react-redux";

export const GameLabels = (props: Props) => {
    return (
        <div>
            <div className="is-size-3 has-text-weight-bold">Player: {props.game?.userName}</div>
            <div className="is-size-3 has-text-weight-bold">Errors: {props.game?.errors}</div>
        </div>
    );
};

const mapStateToProps = (state: { game: Game, highscores: Array<Highscore> }) => {
    return {
        game: state.game
    };
};

export default connect(mapStateToProps)(GameLabels);
