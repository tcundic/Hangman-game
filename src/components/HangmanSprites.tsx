import React from "react";
import {connect} from "react-redux";
import Props from "../models/Props";
import Game from "../models/game";
import Highscore from "../models/highscore";
import Sprite from "../utils/Sprite";

export class HangmanSprites extends React.Component<Props> {

    getSpritePosition = (numberOfErrors: number | undefined) => {
        switch(numberOfErrors) {
            case 0:
                return {x: 0, y: 0, width: 120, height: 170};
            case 1:
                return {x: 120, y: 0, width: 130, height: 170};
            case 2:
                return {x: 250, y: 0, width: 130, height: 170};
            case 3:
                return {x: 370, y: 0, width: 130, height: 170};
            case 4:
                return {x: 0, y: 170, width: 120, height: 170};
            case 5:
                return {x: 120, y: 170, width: 125, height: 170};
            case 6:
                return {x: 240, y: 170, width: 130, height: 170};
            case 7:
                return {x: 370, y: 170, width: 130, height: 170};
            default:
                return {x: 370, y: 170, width: 130, height: 170};
        }
    };

    render() {
        const {x, y, width, height} = this.getSpritePosition(this.props.game?.errors);
        return <Sprite filename="./hangman.png" x={x} y={y} width={width} height={height} />
    }
};

const mapStateToProps = (state: { game: Game, highscores: Array<Highscore> }) => {
    return {
        game: state.game
    };
};

export default connect(mapStateToProps)(HangmanSprites);