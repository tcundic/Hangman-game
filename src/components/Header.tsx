import React from "react";

const Header = () => (
    <div className="header py-2">
        <div className="container">
            <h1 className="has-text-centered">Hangman The Game</h1>
        </div>
    </div>
);

export default Header;