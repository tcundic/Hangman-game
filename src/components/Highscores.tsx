import React, {useEffect} from "react";
import {connect} from "react-redux";
import Props from "../models/Props";
import {mapDispatchToProps, mapStateToProps} from "../utils/utilMethods";
import {getHighscores} from "../utils/http";
import {setHighscores} from "../actions/highscores";
import Highscore from "../models/highscore";
import {useNavigate} from "react-router-dom";

export const Highscores = (props: Props) => {

    const navigate = useNavigate();

    const getNewHighscores = async () => {
        const highscores = (await getHighscores())?.data;

        if (highscores) {
            props.dispatch(setHighscores(highscores));
        }
    }

    useEffect(() => {
        getNewHighscores();
    }, []);

    return (
        <div className="container mt-6">
            <button className="button is-info is-large is-hidden-touch"
                    onClick={() => navigate('/play')}>
                New Game
            </button>
            <button className="button is-info is-large is-hidden-desktop mx-auto is-block-touch mb-4"
                    onClick={() => navigate('/play')}>
                New Game
            </button>
            <h2 className="mb-5 has-text-centered-touch">HIGHSCORES</h2>
            <div className="table-container">
                <table className="table is-striped is-fullwidth mb-6">
                    <thead>
                    <tr>
                        <th>Quote ID</th>
                        <th>Quote Length</th>
                        <th>Unique Characters</th>
                        <th>User Name</th>
                        <th>Number Of Errors</th>
                        <th>Solving Duration</th>
                    </tr>
                    </thead>
                    <tbody>
                    {props.highscores?.map((highscore: Highscore) => {
                        return (
                            <tr>
                                <td>{highscore.quoteId}</td>
                                <td>{highscore.length}</td>
                                <td>{highscore.uniqueCharacters}</td>
                                <td>{highscore.userName}</td>
                                <td>{highscore.errors}</td>
                                <td>{highscore.duration}</td>
                            </tr>
                        )
                    })}
                    </tbody>
                </table>
            </div>
        </div>
    );
};

export default connect(mapStateToProps, mapDispatchToProps)(Highscores);