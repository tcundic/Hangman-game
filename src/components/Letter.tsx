import React, {MouseEventHandler} from "react";

const Letter = ({letter, used, onClick}: { letter: string, used: boolean, onClick: MouseEventHandler }) => {
    return (
        <div className="column">
            {used ?
                <button className="button is-info is-light is-large">{letter}</button> :
                <button onClick={onClick} className="button is-info is-large">{letter}</button>
            }
        </div>
    );
};

export default Letter;