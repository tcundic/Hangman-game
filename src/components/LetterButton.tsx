import React, {MouseEventHandler} from "react";

const LetterButton = ({letter, used, onClick}: { letter: string, used: boolean, onClick: MouseEventHandler }) => {
    return (
        <div className="column is-inline-touch">
            {used ?
                <button className="button is-info is-light is-large my-4">{letter}</button> :
                <button onClick={onClick} className="button is-info is-large my-4">{letter}</button>
            }
        </div>
    );
};

export default LetterButton;