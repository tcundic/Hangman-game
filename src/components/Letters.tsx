import React from "react";
import {connect} from "react-redux";
import Props from "../models/Props";
import Game from "../models/game";
import Highscore from "../models/highscore";
import Letter from "./Letter";
import {useLetter} from "../actions/games";

export class Letters extends React.Component<Props> {
    letters: Array<string>;

    constructor(props: Props) {
        super(props);

        const alpha: Array<number> = Array.from(Array(26)).map((e, i) => i + 65);
        this.letters = alpha.map((x) => String.fromCharCode(x));
    }

    onLetterClick = (e: React.ChangeEvent<any>) => {
        this.props.dispatch(useLetter(e.target.textContent));
    };

    render() {
        return (
            <div className="columns">
                {this.letters.map(letter =>
                    <Letter onClick={this.onLetterClick} key={letter} letter={letter} used={!!this.props.game?.usedLetters.includes(letter)} />
                )}
            </div>
        );
    }
}

const mapStateToProps = (state: { game: Game, highscores: Array<Highscore> }) => {
    return {
        game: state.game
    };
};

export default connect(mapStateToProps)(Letters);