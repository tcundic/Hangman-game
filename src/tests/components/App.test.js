import React from "react";
import { shallow } from "enzyme";
import { App } from "../../components/App";

test('should render App correctly', () => {
    const wrapper = shallow(<App />);
    expect(wrapper).toMatchSnapshot();
});

test('should redirect user when username is not empty', () => {
    const wrapper = shallow(<App game={{ userName: 'abc'}} />);
    expect(wrapper).toMatchSnapshot();
});