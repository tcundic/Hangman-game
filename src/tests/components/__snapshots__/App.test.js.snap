// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render App correctly 1`] = `
<div
  className="container mt-6"
>
  <label
    className="label is-large"
  >
    Username
  </label>
  <div
    className="field is-grouped"
  >
    <p
      className="control"
    >
      <input
        className="input is-large"
        onChange={[Function]}
        placeholder="Enter username"
        type="text"
        value=""
      />
    </p>
    <p
      className="control"
    >
      <a
        className="button is-info is-large"
        onClick={[Function]}
      >
        Save username
      </a>
    </p>
  </div>
</div>
`;
